platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/fingerprint/ConfigurationCacheFingerprintChecker.kt

```diff
@@ -386,34 +389,46 @@ class ConfigurationCacheFingerprintChecker(private val host: Host) {
 
     private
     class ProjectInvalidationState {
-        // When true, the project is definitely invalid
-        // When false, validity is not known
+        // When not null, the project is definitely invalid
+        // When null, validity is not known
         private
-        var invalid = false
+        var _invalidationReason: InvalidationReason? = null
 
         private
         val consumedBy = mutableSetOf<ProjectInvalidationState>()
 
         val isInvalid: Boolean
-            get() = invalid
+            get() = _invalidationReason != null
+
+        val invalidationReason
+            get() = _invalidationReason!!
 
         fun consumedBy(consumer: ProjectInvalidationState) {
-            if (invalid) {
-                consumer.invalidate()
+            if (isInvalid) {
+                invalidateConsumer(consumer)
             } else {
                 consumedBy.add(consumer)
             }
         }
 
-        fun invalidate() {
-            if (invalid) {
+        fun invalidate(reason: StructuredMessageBuilder) {
+            invalidate(StructuredMessage.Builder().apply(reason).build())
+        }
+
+        fun invalidate(reason: InvalidationReason) {
+            if (isInvalid) {
                 return
             }
-            invalid = true
-            for (consumer in consumedBy) {
-                consumer.invalidate()
-            }
+            _invalidationReason = reason
+            consumedBy.forEach(this::invalidateConsumer)
             consumedBy.clear()
         }
+
+        private
+        fun invalidateConsumer(consumer: ProjectInvalidationState) {
+            consumer.invalidate {
+                text("project dependency has changed")
```

Comment:
💅 Can we include the identity path of the dependency here? e.g. `project dependency ':b' has changed`

Reply:
> 💅 Can we include the identity path of the dependency here? e.g. `project dependency ':b' has changed`

Good idea, added.